# Testing Different Classifiers on MNIST Handwritten Digits Image Set

Importing Libraries


In [5]:
from os import listdir
from os.path import isfile,join
import numpy as np
import cv2
from skimage.feature import hog as HOG
from sklearn.linear_model import LogisticRegression 
from sklearn.neighbors import KNeighborsClassifier
from sklearn.grid_search import GridSearchCV
from sklearn.neural_network import MLPClassifier
from sklearn.svm import SVC,NuSVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from MLClassifier import *

Function for Reading Images and Extracting HOG Features

In [6]:
def ReadImages(ListName,FolderName,Label):
    global NumberList
    global responseData
    global trainData
    global hog
    global cv2
    global imutils
    global winSize
    ListName= ListName[0:100]
    for image in ListName:
        img = cv2.imread(join(FolderName,image))
        NumberList.append(img)    
        feature = HOG(cv2.cvtColor(img,cv2.COLOR_RGB2GRAY))
        trainData.append(feature.T)
        responseData.append(Label)

Data Preparation

In [7]:
Zero = '/home/sezan92/SklearnMNIST/mnist_png/training/0'
One= '/home/sezan92/SklearnMNIST/mnist_png/training/1'
Two = '/home/sezan92/SklearnMNIST/mnist_png/training/2'
Three = '/home/sezan92/SklearnMNIST/mnist_png/training/3'
Four = '/home/sezan92/SklearnMNIST/mnist_png/training/4'
Five = '/home/sezan92/SklearnMNIST/mnist_png/training/5'
Six = '/home/sezan92/SklearnMNIST/mnist_png/training/6'
Seven = '/home/sezan92/SklearnMNIST/mnist_png/training/7'
Eight = '/home/sezan92/SklearnMNIST/mnist_png/training/8'
Nine = '/home/sezan92/SklearnMNIST/mnist_png/training/9'

trainData = []
responseData = []
NumberList = []
ZeroImages = [ f for f in listdir(Zero) if isfile(join(Zero,f)) ]
OneImages = [ f for f in listdir(One) if isfile(join(One,f)) ]
TwoImages = [ f for f in listdir(Two) if isfile(join(Two,f)) ]
ThreeImages = [ f for f in listdir(Three) if isfile(join(Three,f)) ]
FourImages = [ f for f in listdir(Four) if isfile(join(Four,f)) ]
FiveImages = [ f for f in listdir(Five) if isfile(join(Five,f)) ]
SixImages = [ f for f in listdir(Six) if isfile(join(Six,f)) ]
SevenImages = [ f for f in listdir(Seven) if isfile(join(Seven,f)) ]
EightImages = [ f for f in listdir(Eight) if isfile(join(Eight,f)) ]
NineImages = [ f for f in listdir(Nine) if isfile(join(Nine,f)) ]

ReadImages(ZeroImages,Zero,0)
ReadImages(OneImages,One,1)
ReadImages(TwoImages,Two,2)
ReadImages(ThreeImages,Three,3)
ReadImages(FourImages,Four,4)
ReadImages(FiveImages,Five,5)
ReadImages(SixImages,Six,6)
ReadImages(SevenImages,Seven,7)
ReadImages(EightImages,Eight,8)
ReadImages(NineImages,Nine,9)

X = np.float32(trainData)
y= np.float32(responseData)

/home/sezan92/anaconda2/lib/python2.7/site-packages/skimage/feature/_hog.py:119: skimage_deprecation: Default value of `block_norm`==`L1` is deprecated and will be changed to `L2-Hys` in v0.15
  'be changed to `L2-Hys` in v0.15', skimage_deprecation)


Classifier

In [8]:
ClassifierSelect(X,y,num_labels=10,SVMFlag=True)    

Knn Training...
Knn Score 0.883
Knn  best Params {'n_neighbors': 7, 'weights': 'distance', 'algorithm': 'auto'}
Logistic Regression Training...
LogReg Score 0.786
LogReg  best Params {'penalty': 'l2', 'tol': 0.0001}
Neural Network Training....


/home/sezan92/anaconda2/lib/python2.7/site-packages/sklearn/neural_network/multilayer_perceptron.py:563: ConvergenceWarning: Stochastic Optimizer: Maximum iterations reached and the optimization hasn't converged yet.
  % (), ConvergenceWarning)


NN Score 0.908
NN  best Params {'activation': 'identity', 'learning_rate': 'constant', 'solver': 'adam'}
SVM training. Caution It is slowest to train....
SVM with NuSVC Score 0.929
SVM with NuSVC best Params{'kernel': 'linear', 'nu': 0.10000000000000001}
DTree Training ...
Decision Tree Score 0.673
Decision Tree params {'splitter': 'best', 'criterion': 'entropy'}
Randomforest Training ....
RTrees Score 0.891
RTrees Best Params {'max_features': 'auto', 'n_estimators': 28, 'criterion': 'gini'}


GridSearchCV(cv=10, error_score='raise',
       estimator=NuSVC(cache_size=200, class_weight=None, coef0=0.0,
   decision_function_shape=None, degree=3, gamma='auto', kernel='rbf',
   max_iter=-1, nu=0.5, probability=False, random_state=None,
   shrinking=True, tol=0.001, verbose=False),
       fit_params={}, iid=True, n_jobs=1,
       param_grid={'kernel': ['linear', 'sigmoid', 'rbf'], 'nu': array([ 0.1,  0.2,  0.3,  0.4,  0.5,  0.6,  0.7,  0.8,  0.9])},
       pre_dispatch='2*n_jobs', refit=True, scoring='accuracy', verbose=0)

SVM is winner in this case. With 92.9% Accuracy! I know using Deep Learning we can achieve 99% accuracy. But The objective of this project is to choose best classifier for a given Dataset, not Predicting Accurately . 